trajopt_ifopt/src/cartesian_line_constraint.cpp

```diff
@@ -0,0 +1,222 @@
+                                           Jacobian& jac_block) const
+{
+  if (use_numeric_differentiation)
+  {
+    auto error_calculator = [&](const Eigen::Ref<const Eigen::VectorXd>& x) { return this->CalcValues(x); };
+    Jacobian jac0(6, n_dof_);
+    jac0 = calcForwardNumJac(error_calculator, joint_vals, 1e-5);
+
+    for (int i = 0; i < 6; i++)
+    {
+      for (int j = 0; j < n_dof_; j++)
+      {
+        // Each jac_block will be for a single variable but for all timesteps. Therefore we must index down to the
+        // correct timestep for this variable
+        jac_block.coeffRef(i, j) = jac0.coeffRef(i, j);
+      }
+    }
+  }
+  else
+  {
+    // Reserve enough room in the sparse matrix
+    jac_block.reserve(n_dof_ * 6);
+
+    Eigen::MatrixXd jac0(6, n_dof_);
+    Eigen::Isometry3d tf0;
+
+    // Calculate the jacobian
+    tf0 = kinematic_info_->manip->calcFwdKin(joint_vals);
+    jac0 = kinematic_info_->manip->calcJacobian(joint_vals, kinematic_info_->kin_link->link_name);
+    tesseract_kinematics::jacobianChangeBase(jac0, kinematic_info_->world_to_base);
+    tesseract_kinematics::jacobianChangeRefPoint(
+        jac0,
+        (kinematic_info_->world_to_base * tf0).linear() *
+            (kinematic_info_->kin_link->transform * kinematic_info_->tcp).translation());
+    tesseract_kinematics::jacobianChangeBase(jac0, line_point_inv_);
+
+    // Paper:
+    // https://ethz.ch/content/dam/ethz/special-interest/mavt/robotics-n-intelligent-systems/rsl-dam/documents/RobotDynamics2016/RD2016script.pdf
+    // The jacobian of the robot is the geometric jacobian (Je) which maps generalized velocities in
+    // joint space to time derivatives of the end-effector configuration representation. It does not
+    // represent the analytic jacobian (Ja) given by a partial differentiation of position and rotation
+    // to generalized coordinates. Since the geometric jacobian is unique there exists a linear mapping
+    // between velocities and the derivatives of the representation.
+    //
+    // The approach in the paper was tried but it was having issues with getting correct jacobian.
+    // Must of had an error in the implementation so should revisit at another time but the approach
+    // below should be sufficient and faster than numerical calculations using the err function.
+
+    // The approach below leverages the geometric jacobian and a small step in time to approximate
+    // the partial derivative of the error function. Note that the rotational portion is the only part
+    // that is required to be modified per the paper.
+    Eigen::Isometry3d pose_err = line_point_inv_ * tf0;
+    Eigen::Vector3d rot_err = calcRotationalError(pose_err.rotation());
+    for (int c = 0; c < jac0.cols(); ++c)
+    {
+      auto new_pose_err = addTwist(pose_err, jac0.col(c), 1e-5);
+      Eigen::VectorXd new_rot_err = calcRotationalError(new_pose_err.rotation());
+      jac0.col(c).tail(3) = ((new_rot_err - rot_err) / 1e-5);
+    }
```

Comment:
If you don't have rotation in the error you should comment it in the jacobian as well.